src/mam/mod_mam_meta.erl

```diff
@@ -269,17 +271,30 @@ add_default_rdbms_opts(Opts) ->
                   none -> [DefaultOpt | Acc];
                   _ -> Acc
               end
-      end,
-      Opts,
-      [{cache_users, true}, {async_writer, true}]).
-
+      end, Opts, [{async_writer, true}]).
+
+add_rdbms_cache_opts(Opts) ->
+    case {lists:keyfind(cache_users, 1, Opts), lists:keyfind(cache_config, 1, Opts)} of
+        {{cache_users, false}, _} ->
+            lists:keydelete(cache_config, 1, Opts);
+        {{cache_users, true}, false} ->
+            [{cache_config, []} | Opts];
+        {false, false} ->
+            [{cache_config, []} | Opts];
+        {false, {cache_config, _}} ->
+            Opts
+    end.
 
 -spec parse_rdbms_opt(Type :: pm | muc, module(), module(),
-                        Option :: {module(), term()}, deps()) -> deps().
+                      Option :: {module(), term()}, deps()) -> deps().
 parse_rdbms_opt(Type, ModRDBMSArch, ModAsyncWriter, Option, Deps) ->
     case Option of
-        {cache_users, true} ->
-            add_dep(mod_mam_cache_user, [Type], Deps);
+        {cache_config, CacheOpts} ->
+            Deps1 = case gen_mod:get_opt(cache_module, CacheOpts, internal) of
+                        internal -> Deps;
+                        mod_cache_users -> add_dep(mod_cache_users, CacheOpts, Deps)
```

Comment:
What would happen if `mod_cache_users` is configured on its own with different options?

Reply:
That is a fantastic question, and I can only guess (expect!) that `gen_mod_deps` is then choosing those configured for the actual module, like, mam won't override what mod_cache_users actually chose. But, I don't actually know, will check.
A similar issue we'd have for `mod_blocking` against `mod_privacy` btw, that's why I naively expect that problem to be solve already 😓

Reply:
I checked this both in the code and in the live shell and the options are merged with overriding, but it's difficult to predict which one will win - in my case the ones from `mod_mam_meta` took precedence. It depends on the order of modules, which (as I quickly checked) currently is alphabetical and the last one wins.

```when=2021-11-16T08:29:45.513101+00:00 level=warning what=overriding_argument pid=<0.419.0> at=gen_mod_deps:merge_args/3:154 old_value={time_to_live,2} new_value={time_to_live,42} module=mod_cache_users```

My recommendation is in this case to completely forbid setting the cache config options in case `mod_cache_users` is used instead of `internal` to prevent confusion. This way we would also get rid of the whole extra complexity in `mod_mam_meta`, which is IMO complex enough.

Anyway, the deps are declared twice: once in `mod_mam_cache_user` without options and then again in `mod_mam_meta` with options. Let's just leave the latter and it will be fine.